rust/phoenix-sdk/src/utils.rs

```diff
@@ -0,0 +1,94 @@
+    create_claim_seat_instruction, create_evict_seat_instruction, EvictTraderAccountBackup,
+};
+use solana_program::{instruction::Instruction, pubkey::Pubkey};
+use spl_associated_token_account::{
+    get_associated_token_address, instruction::create_associated_token_account,
+};
+
+pub async fn create_ata_ix_if_needed(
+    client: &EllipsisClient,
+    payer: &Pubkey,
+    trader: &Pubkey,
+    mint: &Pubkey,
+) -> Option<Instruction> {
+    let ata_address = get_associated_token_address(trader, mint);
+    let ata_account = client.get_account(&ata_address).await;
+
+    if let Ok(ata_account) = ata_account {
+        if !ata_account.data.is_empty() {
+            return None;
+        }
+    }
+    Some(create_associated_token_account(
+        payer,
+        trader,
+        mint,
+        &spl_token::ID,
+    ))
+}
+
+// Check if seat already exists, if not, create seat instruction.
+// Check if the market trader state is full, if so, find a seat to evict and add the evict instruction.
+pub async fn create_claim_seat_ix_if_needed(
+    client: &EllipsisClient,
+    market_pubkey: &Pubkey,
+    trader: &Pubkey,
+) -> anyhow::Result<Vec<Instruction>> {
+    let seat_address = get_seat_address(market_pubkey, trader).0;
+    let seat_account = client.get_account(&seat_address).await;
+
+    let mut instructions = vec![];
+
+    let market_bytes = client.get_account_data(market_pubkey).await?;
+    let (header_bytes, market_bytes) = market_bytes.split_at(size_of::<MarketHeader>());
+    let market_header = bytemuck::try_from_bytes::<MarketHeader>(header_bytes)
+        .map_err(|e| anyhow::anyhow!("Error deserializing market header. Error: {:?}", e))?;
+
+    let market =
+        dispatch_market::load_with_dispatch(&market_header.market_size_params, market_bytes)?.inner;
+
+    let trader_tree = market.get_registered_traders();
+    let num_traders = trader_tree.len();
+    let max_traders = trader_tree.capacity();
+
+    if num_traders == max_traders {
+        //Find a seat to evict (a trader with no locked base or quote lots) and evict trader.
+        for (trader_pubkey, trader_state) in trader_tree.iter() {
+            if trader_state.base_lots_locked == 0 && trader_state.quote_lots_locked == 0 {
+                let evict_trader_state = EvictTraderAccountBackup {
+                    trader_pubkey: *trader_pubkey,
+                    base_token_account_backup: None,
+                    quote_token_account_backup: None,
+                };
+                let evict_trader_ix = create_evict_seat_instruction(
+                    market_pubkey,
+                    &market_header.base_params.mint_key,
+                    &market_header.quote_params.mint_key,
+                    trader,
+                    vec![evict_trader_state],
+                );
+                instructions.push(evict_trader_ix);
+                break;
+            }
+        }
+        return Err(anyhow::anyhow!(
```

Comment:
I don't think this logic is correct. You should only return error if there are no traders to evict (which should also just never be the same in the trader state is full)